### ISSUE TITLE
Wishbone bus high latency in read on Vexriscv CPU "minimal" variant and read/write on others CPU.

### ISSUE THREAD

**Issue by mgaggero** [enjoy-digital/litex#1863]
Hello,
    I'm currently working with LiteX on a Muselab IceSugar board equipped with a Lattice iCE40U5k FPGA. I've encountered an unexpectedly high latency when reading from the Wishbone bus. I've implemented a 32-bit counter that increments every clock cycle and can be reset and read.

What I've observed is that two consecutive reads show a difference of approximately 400 clock cycles when using the Vexriscv CPU with the "minimal" variant. Interestingly, the "lite" variant doesn't exhibit this latency. When employing other CPUs such as picorv32 and femtorv, both read and write operations result in a consistent 270 cycles latency.

This latency issue appears to be related to the Wishbone bus, as the Vexriscv32 "minimal" variant, picorv32, and femtorv all exhibit the same behavior.

**Comment by Dolu1990** [comment]
Hi,

What memory system are you using ? some XIP execution from flash ?

**Comment by mgaggero** [comment]
Absolutely, according to my understanding, the code is executed directly from the Quad SPI flash. There is no copying of program data to the SRAM. It's important to highlight that the Vexriscv 'minimal' variant lacks an instruction/data cache.

**Comment by Dolu1990** [comment]
I think the reason why picorv32, and femtorv are faster in that specific case, is because is pipelined => fetch stuff in advance, which endup being wrong sometime (branch missprediction) meaning it will trash some fetch bandwidth,

**Comment by enjoy-digital** [comment]
The key point is that this iCESugar setup executes the BIOS/software XIP from SPI flash. So the benchmark is not only doing one MMIO access: the CPU is also fetching all benchmark instructions through the SPI flash path.

That explains the different CPU results:

  - `vexriscv-minimal`: no I-cache, so instruction fetches around the MMIO read come directly from SPI flash. This gives the worst result, around 400 cycles.
  - `vexriscv-lite`: has an I-cache, so after the code is cached the repeated read loop no longer pays the SPI flash instruction-fetch cost in the same way.
  - `picorv32` / `femtorv`: the ~270 cycle read/write result is still mostly software execution + XIP fetch cost. Their fetch/pipeline behavior is different from VexRiscv minimal, so they can give a
  different number, but this still does not mean the Wishbone read/write handshake itself takes 270 cycles.

So the test is measuring the complete CPU/software/XIP path:

  - instruction fetches from SPI flash,
  - CPU pipeline/load-store behavior,
  - compiler-generated code around the MMIO access,
  - possible function call / loop overhead,
  - and only then the actual Wishbone/CSR transaction.

The LiteX CSR bus itself is much lower latency: CSR writes complete in one CSR-bus cycle and reads in two CSR-bus cycles, plus the small fixed Wishbone-to-CSR bridge overhead.
 To measure the real bus latency, the test should either run from SRAM/internal RAM, or observe `cyc/stb/ack` directly in simulation/ILA, or use a small hardware-side monitor that counts cycles between Wishbone request and acknowledge.

So I don’t think this shows a generic Wishbone latency issue. It is mainly an effect of running the benchmark from SPI flash, with the exact number depending on the CPU implementation/cache/prefetch behavior. I think we can close unless there is a current reproducer showing the Wishbone `ack` itself taking hundreds of cycles at the bus-signal level.
